typing/printtyp.ml

```diff
@@ -1721,16 +1721,35 @@ let rec trace fst txt ppf = function
        (trace false txt) rem
   | _ -> ()
 
-let rec filter_trace keep_last = function
-  | Trace.(Diff ({ got=t1, t1'; expected=t2, t2'} as elt)) :: rem ->
-      let rem' = filter_trace keep_last rem in
-      if is_constr_row ~allow_ident:true t1'
+let is_discarded is_last = function
+  | Trace.(Diff { got=t1, t1'; expected=t2, t2'}) ->
+      is_constr_row ~allow_ident:true t1'
       || is_constr_row ~allow_ident:true t2'
-      || same_path t1 t1' && same_path t2 t2' && not (keep_last && rem' = [])
-      then rem'
-      else elt :: rem'
-  | _elt :: rem -> filter_trace keep_last rem
-  | _ -> []
+      || same_path t1 t1' && same_path t2 t2' && not is_last
+  | _ -> false
+
+(** Flatten the trace and remove elements that are always discarded
+    during printing *)
+let prepare_trace f tr =
+  let rec clean_trace = function
+    | [] -> []
+    | x :: q -> let rem = clean_trace q in
+        if is_discarded (rem=[]) x then rem
+        else x :: rem
+  in
+  match Trace.flatten f tr with
+  | [] -> []
+  | elt :: rem -> (* the first element is always kept *)
+      elt :: clean_trace rem
+
+(** Keep elements that are not [Diff _ ] and take the decision
+    for the last element, require a prepared trace *)
+let rec filter_trace keep_last = function
+  | [] -> []
+  | [Trace.Diff d as elt] when is_discarded false elt ->
```

Comment:
I find `is_discarded false elt` confusing. The name of the first argument of `is_discarded` is `is_last`, and you set it to `false`, but `elt` *is* the last element! At this point I'm a bit too confused to understand whether this line and the next one are actually correct, but I would think something like follows is more natural:

```ocaml
| [Trace.Diff d as elt] ->
  if is_discarded keep_last elt then [] else [d]
```

Reply:
Maybe a more natural solution would be to make `is_discarded` returns a variant `Discard | Keep_if_last |  Keep` then the previous line would be checking `when is_discarded elt = Keep_if_last` .

Reply:
I went ahead with this solution.